modules/lightmapper_rd/lightmapper_rd.cpp

```diff
@@ -671,6 +672,114 @@ LightmapperRD::BakeError LightmapperRD::_dilate(RenderingDevice *rd, Ref<RDShade
+	} else {
+		lib_path = oidn_lib_path.path_join("libOpenImageDenoise.so");
+	}
+
+	_unload_oidn();
+
+	if (OS::get_singleton()->open_dynamic_library(lib_path, oidn_lib_handle, true) != OK) {
+		oidn_lib_handle = nullptr;
+		ERR_PRINT(vformat("Failed to load %s.", lib_path));
+		return false;
+	}
+	bool symbols_ok = true;
+	symbols_ok = symbols_ok && (OS::get_singleton()->get_dynamic_library_symbol_handle(oidn_lib_handle, "oidnNewDevice", (void *&)oidnNewDevice) == OK);
+	symbols_ok = symbols_ok && (OS::get_singleton()->get_dynamic_library_symbol_handle(oidn_lib_handle, "oidnCommitDevice", (void *&)oidnCommitDevice) == OK);
+	symbols_ok = symbols_ok && (OS::get_singleton()->get_dynamic_library_symbol_handle(oidn_lib_handle, "oidnNewFilter", (void *&)oidnNewFilter) == OK);
+	symbols_ok = symbols_ok && (OS::get_singleton()->get_dynamic_library_symbol_handle(oidn_lib_handle, "oidnSetSharedFilterImage", (void *&)oidnSetSharedFilterImage) == OK);
+	symbols_ok = symbols_ok && (OS::get_singleton()->get_dynamic_library_symbol_handle(oidn_lib_handle, "oidnSetFilterBool", (void *&)oidnSetFilterBool) == OK);
+	symbols_ok = symbols_ok && (OS::get_singleton()->get_dynamic_library_symbol_handle(oidn_lib_handle, "oidnCommitFilter", (void *&)oidnCommitFilter) == OK);
+	symbols_ok = symbols_ok && (OS::get_singleton()->get_dynamic_library_symbol_handle(oidn_lib_handle, "oidnExecuteFilter", (void *&)oidnExecuteFilter) == OK);
+	symbols_ok = symbols_ok && (OS::get_singleton()->get_dynamic_library_symbol_handle(oidn_lib_handle, "oidnGetDeviceError", (void *&)oidnGetDeviceError) == OK);
+	symbols_ok = symbols_ok && (OS::get_singleton()->get_dynamic_library_symbol_handle(oidn_lib_handle, "oidnReleaseFilter", (void *&)oidnReleaseFilter) == OK);
+	symbols_ok = symbols_ok && (OS::get_singleton()->get_dynamic_library_symbol_handle(oidn_lib_handle, "oidnReleaseDevice", (void *&)oidnReleaseDevice) == OK);
+	if (!symbols_ok) {
+		ERR_PRINT("Failed to load OIDN symbols.");
+		_unload_oidn();
+		return false;
+	}
+
+	oidn_device = oidnNewDevice(OIDN_DEVICE_TYPE_CPU);
+	oidnCommitDevice(oidn_device);
+	const char *msg = nullptr;
+	if (oidnGetDeviceError(oidn_device, &msg) != OIDN_ERROR_NONE) {
+		ERR_PRINT("Failed to load OIDN device.");
+		_unload_oidn();
+		return false;
+	}
+
+	return true;
+}
+
+void LightmapperRD::_unload_oidn() {
+	if (oidn_lib_handle) {
+		if (oidn_device) {
+			oidnReleaseDevice(oidn_device);
+			oidn_device = nullptr;
+		}
+
+		OS::get_singleton()->close_dynamic_library(oidn_lib_handle);
+		oidn_lib_handle = nullptr;
+	}
+}
+
+LightmapperRD::BakeError LightmapperRD::_denoise_oidn(RenderingDevice *p_rd, RID p_source_light_tex, RID p_source_normal_tex, RID p_dest_light_tex, const Size2i &p_atlas_size, int p_atlas_slices, bool p_bake_sh) {
+	for (int i = 0; i < p_atlas_slices * (p_bake_sh ? 4 : 1); i++) {
+		Vector<uint8_t> sl = p_rd->texture_get_data(p_source_light_tex, i);
+		Vector<uint8_t> sn = p_rd->texture_get_data(p_source_normal_tex, i);
```

Comment:
I think this might not work with spherical harmonics (the "Directional" toggle, p_bake_sh) because the normal texture will have less slices than the light texture (which has 4 times the slices when SH are used).

If that is the case, you can probably fix it by just switching to a double for loop: 0 to < p_atlas_slices, fetch the normal texture data, then 0 to < 4 for each light slice.

Reply:
This seems to be the case.

Note: updated PRs, but I'm not entirely sure which order light slices come in, so will need extra checking.

Reply:
You got the order correct as far as I can tell.

Reply:
Tested with more complex scene, order seems to be correct.

Reply:
@bruvzg do you have instructions to test this?